File: recyclerview/src/androidTest/java/com/kfaraj/support/widget/SupportRecyclerViewTest.java
package com.kfaraj.support.widget;

import android.content.Context;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.MeasureSpec;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.test.InstrumentationRegistry;
import androidx.test.filters.SmallTest;
import androidx.test.runner.AndroidJUnit4;

import com.kfaraj.support.widget.SupportRecyclerView.MultiChoiceModeListener;
import com.kfaraj.support.widget.SupportRecyclerView.OnItemClickListener;
import com.kfaraj.support.widget.SupportRecyclerView.OnItemLongClickListener;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Collections;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class SupportRecyclerViewTest {

    private SupportRecyclerView mRecyclerView;
    private View mEmptyView;
    private MockAdapter mAdapter;
    private MockOnItemClickListener mOnItemClickListener;
    private MockOnItemLongClickListener mOnItemLongClickListener;
    private MockMultiChoiceModeListener mMultiChoiceModeListener;

    @Before
    public void setUp() {
        final Context context = InstrumentationRegistry.getContext();
        mRecyclerView = new SupportRecyclerView(context);
        mEmptyView = new View(context);
        mAdapter = new MockAdapter();
        mOnItemClickListener = new MockOnItemClickListener();
        mOnItemLongClickListener = new MockOnItemLongClickListener();
        mMultiChoiceModeListener = new MockMultiChoiceModeListener();
        mRecyclerView.setLayoutManager(new LinearLayoutManager(context));
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.setEmptyView(mEmptyView);
        mRecyclerView.setOnItemClickListener(mOnItemClickListener);
        mRecyclerView.setOnItemLongClickListener(mOnItemLongClickListener);
        mRecyclerView.setMultiChoiceModeListener(mMultiChoiceModeListener);
    }

    @Test
    public void testEmptyView() {
        assertEquals(mEmptyView, mRecyclerView.getEmptyView());
        populateAdapter(0);
        assertEquals(View.VISIBLE, mEmptyView.getVisibility());
        populateAdapter(1);
        assertEquals(View.GONE, mEmptyView.getVisibility());
    }

    @Test
    public void testOnItemClickListener() {
        assertEquals(mOnItemClickListener, mRecyclerView.getOnItemClickListener());
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performClick();
        assertTrue(mOnItemClickListener.called);
    }

    @Test
    public void testOnItemLongClickListener() {
        assertEquals(mOnItemLongClickListener, mRecyclerView.getOnItemLongClickListener());
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performLongClick();
        assertTrue(mOnItemLongClickListener.called);
    }

    @Test
    public void testChoiceMode() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_NONE);
        assertEquals(SupportRecyclerView.CHOICE_MODE_NONE, mRecyclerView.getChoiceMode());
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_SINGLE);
        assertEquals(SupportRecyclerView.CHOICE_MODE_SINGLE, mRecyclerView.getChoiceMode());
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        assertEquals(SupportRecyclerView.CHOICE_MODE_MULTIPLE, mRecyclerView.getChoiceMode());
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL);
        assertEquals(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL, mRecyclerView.getChoiceMode());
    }

    @Test
    public void testItemClick_choiceModeNone() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_NONE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performClick();
        assertFalse(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemClickListener.called);
    }

    @Test
    public void testItemClick_choiceModeSingle() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_SINGLE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performClick();
        assertTrue(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemClickListener.called);
    }

    @Test
    public void testItemClick_choiceModeMultiple() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performClick();
        assertTrue(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemClickListener.called);
    }

    @Test
    public void testItemClick_choiceModeMultipleModal() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performClick();
        assertFalse(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemClickListener.called);
    }

    @Test
    public void testItemLongClick_choiceModeNone() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_NONE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performLongClick();
        assertFalse(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemLongClickListener.called);
    }

    @Test
    public void testItemLongClick_choiceModeSingle() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_SINGLE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performLongClick();
        assertFalse(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemLongClickListener.called);
    }

    @Test
    public void testItemLongClick_choiceModeMultiple() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performLongClick();
        assertFalse(mRecyclerView.isItemChecked(0));
        assertTrue(mOnItemLongClickListener.called);
    }

    @Test
    public void testItemLongClick_choiceModeMultipleModal() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL);
        populateAdapter(1);
        mRecyclerView.getChildAt(0).performLongClick();
        assertTrue(mRecyclerView.isItemChecked(0));
        assertFalse(mOnItemLongClickListener.called);
    }

    @Test
    public void testItemChecked_choiceModeNone() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_NONE);
        populateAdapter(1);
        mRecyclerView.setItemChecked(0, true);
        assertEquals(0, mRecyclerView.getCheckedItemCount());
    }

    @Test
    public void testItemChecked_choiceModeSingle() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_SINGLE);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, true);
        mRecyclerView.setItemChecked(1, true);
        assertEquals(1, mRecyclerView.getCheckedItemCount());
        assertEquals(1, mRecyclerView.getCheckedItemPosition());
        mRecyclerView.clearChoices();
        assertEquals(0, mRecyclerView.getCheckedItemCount());
    }

    @Test
    public void testItemChecked_choiceModeMultiple() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, true);
        mRecyclerView.setItemChecked(1, true);
        assertEquals(2, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(0));
        assertTrue(mRecyclerView.isItemChecked(1));
        mRecyclerView.clearChoices();
        assertEquals(0, mRecyclerView.getCheckedItemCount());
    }

    @Test
    public void testItemChecked_choiceModeMultipleModal() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, true);
        mRecyclerView.setItemChecked(1, true);
        assertEquals(2, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(0));
        assertTrue(mRecyclerView.isItemChecked(1));
        mRecyclerView.clearChoices();
        assertEquals(0, mRecyclerView.getCheckedItemCount());
    }

    @Test
    public void testItemChecked_adapterChange() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, false);
        mRecyclerView.setItemChecked(1, true);
        mAdapter.getItems().remove(0);
        mAdapter.getItems().add(0, new Object());
        Collections.swap(mAdapter.getItems(), 0, 1);
        mAdapter.notifyDataSetChanged();
        assertEquals(1, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(0));
    }

    @Test
    public void testItemChecked_adapterInsert() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(1);
        mRecyclerView.setItemChecked(0, true);
        mAdapter.getItems().add(0, new Object());
        mAdapter.notifyItemInserted(0);
        assertEquals(1, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(1));
    }

    @Test
    public void testItemChecked_adapterRemove() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, false);
        mRecyclerView.setItemChecked(1, true);
        mAdapter.getItems().remove(0);
        mAdapter.notifyItemRemoved(0);
        assertEquals(1, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(0));
    }

    @Test
    public void testItemChecked_adapterMove() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(2);
        mRecyclerView.setItemChecked(0, false);
        mRecyclerView.setItemChecked(1, true);
        Collections.swap(mAdapter.getItems(), 0, 1);
        mAdapter.notifyItemMoved(0, 1);
        assertEquals(1, mRecyclerView.getCheckedItemCount());
        assertTrue(mRecyclerView.isItemChecked(0));
    }

    @Test
    public void testCheckedItemPositions() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(10);
        for (int i = 0; i < 10; i++) {
            mRecyclerView.setItemChecked(i, i % 2 == 0);
        }
        SparseBooleanArray positions = mRecyclerView.getCheckedItemPositions();
        for (int i = 0; i < positions.size(); i++) {
            int position = positions.keyAt(i);
            boolean checked = positions.valueAt(i);
            assertEquals(mRecyclerView.isItemChecked(position), checked);
        }
    }

    @Test
    public void testCheckedItemIds() {
        mRecyclerView.setChoiceMode(SupportRecyclerView.CHOICE_MODE_MULTIPLE);
        populateAdapter(10);
        for (int i = 0; i < 10; i++) {
            mRecyclerView.setItemChecked(i, i % 2 == 0);
        }
        long[] ids = mRecyclerView.getCheckedItemIds();
        assertEquals(mRecyclerView.getCheckedItemCount(), ids.length);
    }

    @Test
    public void testMultiChoiceModeListener() {
        assertEquals(mMultiChoiceModeListener, mRecyclerView.getMultiChoiceModeListener());
    }

    private void populateAdapter(int count) {
        mAdapter.getItems().clear();
        for (int i = 0; i < count; i++) {
            mAdapter.getItems().add(new Object());
        }
        mAdapter.notifyDataSetChanged();
        requestLayout();
    }

    private void requestLayout() {
        final int widthMeasureSpec = MeasureSpec.makeMeasureSpec(320, MeasureSpec.EXACTLY);
        final int heightMeasureSpec = MeasureSpec.makeMeasureSpec(320, MeasureSpec.EXACTLY);
        mRecyclerView.measure(widthMeasureSpec, heightMeasureSpec);
        mRecyclerView.layout(0, 0, 320, 320);
    }

    private static class MockAdapter extends RecyclerView.Adapter<MockViewHolder> {

        private final ArrayList<Object> mItems = new ArrayList<>();

        MockAdapter() {
            setHasStableIds(true);
        }

        ArrayList<Object> getItems() {
            return mItems;
        }

        @NonNull
        @Override
        public MockViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View itemView = new View(parent.getContext());
            return new MockViewHolder(itemView);
        }

        @Override
        public void onBindViewHolder(@NonNull MockViewHolder holder, int position) {
        }

        @Override
        public long getItemId(int position) {
            return mItems.get(position).hashCode();
        }

        @Override
        public int getItemCount() {
            return mItems.size();
        }

    }

    private static class MockViewHolder extends RecyclerView.ViewHolder {

        MockViewHolder(View itemView) {
            super(itemView);
        }

    }

    private static class MockOnItemClickListener implements OnItemClickListener {

        boolean called = false;

        @Override
        public void onItemClick(SupportRecyclerView parent,
                View view, int position, long id) {
            this.called = true;
        }

    }

    private static class MockOnItemLongClickListener implements OnItemLongClickListener {

        boolean called = false;

        @Override
        public boolean onItemLongClick(SupportRecyclerView parent,
                View view, int position, long id) {
            this.called = true;
            return true;
        }

    }

    private static class MockMultiChoiceModeListener implements MultiChoiceModeListener {

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return true;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            return true;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
        }

        @Override
        public void onItemCheckedStateChanged(ActionMode mode,
                int position, long id, boolean checked) {
        }

    }

}

File: recyclerview/src/androidTest/java/com/kfaraj/support/widget/SupportRecyclerViewTest.kt
package com.kfaraj.support.widget

import android.content.Context
import android.view.View
import android.view.View.MeasureSpec
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.filters.SmallTest
import com.kfaraj.support.testutils.TestAdapter
import com.kfaraj.support.testutils.TestMultiChoiceModeListener
import com.kfaraj.support.testutils.TestOnItemClickListener
import com.kfaraj.support.testutils.TestOnItemLongClickListener
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.util.*

@SmallTest
class SupportRecyclerViewTest {

    private lateinit var recyclerView: SupportRecyclerView
    private lateinit var emptyView: View
    private lateinit var adapter: TestAdapter<Any>
    private lateinit var onItemClickListener: TestOnItemClickListener
    private lateinit var onItemLongClickListener: TestOnItemLongClickListener
    private lateinit var multiChoiceModeListener: TestMultiChoiceModeListener

    @Before
    fun setUp() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        recyclerView = SupportRecyclerView(context)
        emptyView = View(context)
        adapter = TestAdapter()
        onItemClickListener = TestOnItemClickListener()
        onItemLongClickListener = TestOnItemLongClickListener()
        multiChoiceModeListener = TestMultiChoiceModeListener()
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = adapter
        recyclerView.emptyView = emptyView
        recyclerView.onItemClickListener = onItemClickListener
        recyclerView.onItemLongClickListener = onItemLongClickListener
        recyclerView.multiChoiceModeListener = multiChoiceModeListener
    }

    @Test
    fun testEmptyView() {
        assertEquals(emptyView, recyclerView.emptyView)
        populateAdapter(0)
        assertEquals(View.VISIBLE, emptyView.visibility)
        populateAdapter(1)
        assertEquals(View.GONE, emptyView.visibility)
    }

    @Test
    fun testOnItemClickListener() {
        assertEquals(onItemClickListener, recyclerView.onItemClickListener)
        populateAdapter(1)
        recyclerView.getChildAt(0).performClick()
        assertTrue(onItemClickListener.invoked)
    }

    @Test
    fun testOnItemLongClickListener() {
        assertEquals(onItemLongClickListener, recyclerView.onItemLongClickListener)
        populateAdapter(1)
        recyclerView.getChildAt(0).performLongClick()
        assertTrue(onItemLongClickListener.invoked)
    }

    @Test
    fun testChoiceMode() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_NONE
        assertEquals(SupportRecyclerView.CHOICE_MODE_NONE, recyclerView.choiceMode)
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_SINGLE
        assertEquals(SupportRecyclerView.CHOICE_MODE_SINGLE, recyclerView.choiceMode)
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        assertEquals(SupportRecyclerView.CHOICE_MODE_MULTIPLE, recyclerView.choiceMode)
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL
        assertEquals(SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL, recyclerView.choiceMode)
    }

    @Test
    fun testMultiChoiceModeListener() {
        assertEquals(multiChoiceModeListener, recyclerView.multiChoiceModeListener)
    }

    @Test
    fun testItemClick_choiceModeNone() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_NONE
        populateAdapter(1)
        recyclerView.getChildAt(0).performClick()
        assertFalse(recyclerView.isItemChecked(0))
        assertTrue(onItemClickListener.invoked)
    }

    @Test
    fun testItemClick_choiceModeSingle() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_SINGLE
        populateAdapter(1)
        recyclerView.getChildAt(0).performClick()
        assertTrue(recyclerView.isItemChecked(0))
        assertTrue(onItemClickListener.invoked)
    }

    @Test
    fun testItemClick_choiceModeMultiple() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(1)
        recyclerView.getChildAt(0).performClick()
        assertTrue(recyclerView.isItemChecked(0))
        assertTrue(onItemClickListener.invoked)
    }

    @Test
    fun testItemClick_choiceModeMultipleModal() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL
        populateAdapter(1)
        recyclerView.getChildAt(0).performClick()
        assertFalse(recyclerView.isItemChecked(0))
        assertTrue(onItemClickListener.invoked)
    }

    @Test
    fun testItemLongClick_choiceModeNone() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_NONE
        populateAdapter(1)
        recyclerView.getChildAt(0).performLongClick()
        assertFalse(recyclerView.isItemChecked(0))
        assertTrue(onItemLongClickListener.invoked)
    }

    @Test
    fun testItemLongClick_choiceModeSingle() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_SINGLE
        populateAdapter(1)
        recyclerView.getChildAt(0).performLongClick()
        assertFalse(recyclerView.isItemChecked(0))
        assertTrue(onItemLongClickListener.invoked)
    }

    @Test
    fun testItemLongClick_choiceModeMultiple() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(1)
        recyclerView.getChildAt(0).performLongClick()
        assertFalse(recyclerView.isItemChecked(0))
        assertTrue(onItemLongClickListener.invoked)
    }

    @Test
    fun testItemLongClick_choiceModeMultipleModal() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL
        populateAdapter(1)
        recyclerView.getChildAt(0).performLongClick()
        assertTrue(recyclerView.isItemChecked(0))
        assertFalse(multiChoiceModeListener.invoked)
    }

    @Test
    fun testItemChecked_choiceModeNone() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_NONE
        populateAdapter(1)
        recyclerView.setItemChecked(0, true)
        assertEquals(0, recyclerView.checkedItemCount)
    }

    @Test
    fun testItemChecked_choiceModeSingle() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_SINGLE
        populateAdapter(2)
        recyclerView.setItemChecked(0, true)
        recyclerView.setItemChecked(1, true)
        assertEquals(1, recyclerView.checkedItemCount)
        assertEquals(1, recyclerView.checkedItemPosition)
        recyclerView.clearChoices()
        assertEquals(0, recyclerView.checkedItemCount)
    }

    @Test
    fun testItemChecked_choiceModeMultiple() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(2)
        recyclerView.setItemChecked(0, true)
        recyclerView.setItemChecked(1, true)
        assertEquals(2, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(0))
        assertTrue(recyclerView.isItemChecked(1))
        recyclerView.clearChoices()
        assertEquals(0, recyclerView.checkedItemCount)
    }

    @Test
    fun testItemChecked_choiceModeMultipleModal() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE_MODAL
        populateAdapter(2)
        recyclerView.setItemChecked(0, true)
        recyclerView.setItemChecked(1, true)
        assertEquals(2, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(0))
        assertTrue(recyclerView.isItemChecked(1))
        recyclerView.clearChoices()
        assertEquals(0, recyclerView.checkedItemCount)
    }

    @Test
    fun testItemChecked_adapterChange() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(2)
        recyclerView.setItemChecked(0, false)
        recyclerView.setItemChecked(1, true)
        adapter.items.removeAt(0)
        adapter.items.add(0, Any())
        Collections.swap(adapter.items, 0, 1)
        adapter.notifyDataSetChanged()
        assertEquals(1, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(0))
    }

    @Test
    fun testItemChecked_adapterInsert() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(1)
        recyclerView.setItemChecked(0, true)
        adapter.items.add(0, Any())
        adapter.notifyItemInserted(0)
        assertEquals(1, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(1))
    }

    @Test
    fun testItemChecked_adapterRemove() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(2)
        recyclerView.setItemChecked(0, false)
        recyclerView.setItemChecked(1, true)
        adapter.items.removeAt(0)
        adapter.notifyItemRemoved(0)
        assertEquals(1, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(0))
    }

    @Test
    fun testItemChecked_adapterMove() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(2)
        recyclerView.setItemChecked(0, false)
        recyclerView.setItemChecked(1, true)
        Collections.swap(adapter.items, 0, 1)
        adapter.notifyItemMoved(0, 1)
        assertEquals(1, recyclerView.checkedItemCount)
        assertTrue(recyclerView.isItemChecked(0))
    }

    @Test
    fun testCheckedItemPositions() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(10)
        for (i in 0..9) {
            recyclerView.setItemChecked(i, i % 2 == 0)
        }
        val positions = recyclerView.checkedItemPositions
        assertEquals(recyclerView.checkedItemCount, positions.size())
    }

    @Test
    fun testCheckedItemIds() {
        recyclerView.choiceMode = SupportRecyclerView.CHOICE_MODE_MULTIPLE
        populateAdapter(10)
        for (i in 0..9) {
            recyclerView.setItemChecked(i, i % 2 == 0)
        }
        val ids = recyclerView.checkedItemIds
        assertEquals(recyclerView.checkedItemCount, ids.size)
    }

    private fun populateAdapter(count: Int) {
        adapter.items.clear()
        for (i in 0 until count) {
            adapter.items.add(Any())
        }
        adapter.notifyDataSetChanged()
        requestLayout()
    }

    private fun requestLayout() {
        val widthMeasureSpec = MeasureSpec.makeMeasureSpec(320, MeasureSpec.EXACTLY)
        val heightMeasureSpec = MeasureSpec.makeMeasureSpec(320, MeasureSpec.EXACTLY)
        recyclerView.measure(widthMeasureSpec, heightMeasureSpec)
        recyclerView.layout(0, 0, 320, 320)
    }

}
